File: cpp/src/solutions/ransom_note.cpp
// LeetCode 383. Ransom Note in C++

#include "cpp/src/common/common.h"

class Solution {
 public:
  bool canConstruct(string ransomNote, string magazine) {
    vector<int> flag(26, 0);
    for_each(ransomNote.begin(), ransomNote.end(),
             [&](char &ch) { flag[ch - 'a'] += 1; });
    for_each(magazine.begin(), magazine.end(),
             [&](char &ch) { flag[ch - 'a'] -= 1; });
    auto valid = count_if(flag.begin(), flag.end(),
                          [](int &value) { return value > 0; });
    return valid == 0;
  }
};

TEST(canConstruct, example1) {
  string ransomNote = "a", magazine = "b";
  bool ans = false;
  EXPECT_EQ(Solution().canConstruct(ransomNote, magazine), ans);
}

TEST(canConstruct, example2) {
  string ransomNote = "aa", magazine = "ab";
  bool ans = false;
  EXPECT_EQ(Solution().canConstruct(ransomNote, magazine), ans);
}

TEST(canConstruct, example3) {
  string ransomNote = "aa", magazine = "aab";
  bool ans = true;
  EXPECT_EQ(Solution().canConstruct(ransomNote, magazine), ans);
}

File: rust/src/solutions/ransom_note.rs
// LeetCode 383. Ransom Note

use super::Solution;

impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut flag = vec![0; 26];
        for ch in ransom_note.bytes() {
            flag[ch as usize - 'a' as usize] += 1;
        }
        for ch in magazine.bytes() {
            flag[ch as usize - 'a' as usize] -= 1;
        }
        flag.iter().filter(|&x| *x > 0).count() == 0
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let (ransom_note, magazine) = (String::from("a"), String::from("b"));
        let ans = false;
        assert_eq!(Solution::can_construct(ransom_note, magazine), ans);
    }

    #[test]
    fn example2() {
        let (ransom_note, magazine) = (String::from("aa"), String::from("ab"));
        let ans = false;
        assert_eq!(Solution::can_construct(ransom_note, magazine), ans);
    }

    #[test]
    fn example3() {
        let (ransom_note, magazine) = (String::from("aa"), String::from("aab"));
        let ans = true;
        assert_eq!(Solution::can_construct(ransom_note, magazine), ans);
    }
}
